File: python/heap.py
import util

# sort implements heap sort algorithm. The function takes an array.
def sort(a):
	n = len(a)
	a.insert(0, 0)
	for i in range(n//2, 0, -1):
		sink(a, i, n)
	for i in range(1, n+1):
		util.swap(a, 1, n-i+1)
		sink(a, 1, n-i)
	a.pop(0)

def sink(a, i, n):
	lc = 2*i
	if lc > n:
		return
	rc = lc+1
	mc = 0
	if rc > n:
		mc = lc
	elif a[lc] > a[rc]:
		mc = lc
	else:
		mc = rc
	if a[i] >= a[mc]:
		return
	util.swap(a, i, mc)
	sink(a, mc, n)

File: php/heap.php
<?php
require_once('util.php');

// heap implements heap sort algorithm. The function takes an array.
function heap(&$a) {
	$n = count($a);
	array_unshift($a, 0);
	for($i = $n/2; $i >= 1; $i--)
		sink($a, $i, $n);
	for($i = 1; $i <= $n; $i++) {
		swap($a, 1, $n-$i+1);
		sink($a, 1, $n-$i);
	}
	array_shift($a);
}

function sink(&$a, $i, $n) {
	$lc = 2*$i;
	if($lc > $n)
		return;
	$rc = $lc+1;
	$mc = ($rc > $n) ? $lc : ($a[$lc] > $a[$rc] ? $lc : $rc);
	if($a[$i] >= $a[$mc])
		return;
	swap($a, $i, $mc);
	sink($a, $mc, $n);
}
